File: atcoder/educational_dp/a_frog1.cpp
#include<bits/stdc++.h>

using namespace std;
typedef long long ll;
const ll INF =  1LL << 60;

ll solve (int n, vector<int> h) {
  vector<ll> costs(n, INF);

  costs.at(0) = 0;
  for (int i = 0; i < n; i++) {
    for (int step = 1; step <= 2 && i + step < n; step++) {
      costs.at(i + step) = min(costs.at(i + step), costs.at(i) + abs(h.at(i + step) - h.at(i)));
    }
  }

  int result = costs.at(n-1);
  return result;
}

int main (void) {
  int n;
  scanf("%d", &n);

  vector<int> h(n);
  for (int i = 0; i < n; i++) {
    scanf("%d", &h.at(i));
  }

  ll result = solve(n, h);
  printf("%lld\n", result);

  return 0;
}

File: atcoder/educational_dp/4/a_frog1.py
from typing import List

def solve (N: int, h: List[int]) -> int:

    dp = [float('inf')] * N
    dp[0] = 0
    for i in range(1, N):
        dp[i] = min(
                    [dp[i]] + 
                    [x + abs(h[i] - h[max(0, i - 2) + j]) for j, x in enumerate(dp[max(0, i - 2): i])]
                )

    return dp[N-1]

def main () -> None:
    N = int(input())
    h = [int(x) for x in input().split()]

    result = solve(N, h)

    print(result)

    return

if __name__ == '__main__':
    main()
